{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```bash\n",
    "# We need to combine a few datasets:\n",
    "#\n",
    "# * ICU (work with Paul Wischmeyer).\n",
    "# * Infant time series (work with Ruth Ley).\n",
    "# * Fecal Microbiome Transplant (work with Mike Sadowsky).\n",
    "# * American Gut Data.\n",
    "#\n",
    "#\n",
    "# While seemingly trivial, this requires some finesse to properly combine\n",
    "# the data.\n",
    "# 1) We need to start by deblurring the ICU and FMT data, so first we need\n",
    "#    to grab the data from Qiita (HDF5 formatted sequences) and import them\n",
    "#    to QIIME2.\n",
    "#\n",
    "mkdir -p demux-files\n",
    "cp /projects/qiita_data/preprocessed_data/175_seqs.demux /home/yovazquezbaeza/research/ag-animation/demux-files/fmt-seqs.demux\n",
    "cp /projects/qiita_data/preprocessed_data/615_seqs.demux /home/yovazquezbaeza/research/ag-animation/demux-files/icu-seqs.demux\n",
    "#\n",
    "~/bin/make_importable.sh /home/yovazquezbaeza/research/ag-animation/demux-files/fmt-seqs.demux /home/yovazquezbaeza/research/ag-animation/fmt-sequences-qiime2 1\n",
    "qiime tools import --input-path /home/yovazquezbaeza/research/ag-animation/fmt-sequences-qiime2/ --type 'SampleData[SequencesWithQuality]' --output-path /home/yovazquezbaeza/research/ag-animation/fmt-seqs.qza\n",
    "#\n",
    "~/bin/make_importable.sh /home/yovazquezbaeza/research/ag-animation/demux-files/icu-seqs.demux /home/yovazquezbaeza/research/ag-animation/icu-sequences-qiime2 1\n",
    "qiime tools import --input-path /home/yovazquezbaeza/research/ag-animation/icu-sequences-qiime2/ --type 'SampleData[SequencesWithQuality]' --output-path /home/yovazquezbaeza/research/ag-animation/icu-seqs.qza\n",
    "#\n",
    "# 1.1) Deblur the sequences and select trimming at 125 NT.\n",
    "#\n",
    "qiime deblur denoise-16S --output-dir /home/yovazquezbaeza/research/ag-animation/deblur-fmt --i-demultiplexed-seqs /home/yovazquezbaeza/research/ag-animation/fmt-seqs.qza --p-trim-length 125 --p-no-hashed-feature-ids --verbose --p-jobs-to-start 16\n",
    "qiime deblur denoise-16S --output-dir /home/yovazquezbaeza/research/ag-animation/deblur-icu --i-demultiplexed-seqs /home/yovazquezbaeza/research/ag-animation/icu-seqs.qza --p-trim-length 125 --p-no-hashed-feature-ids --verbose --p-jobs-to-start 16\n",
    "#\n",
    "# 1.2) Merge sequences\n",
    "#\n",
    "qiime feature-table merge --i-table1 /home/yovazquezbaeza/research/ag-animation/deblur-fmt/table.qza --i-table2 /home/yovazquezbaeza/research/ag-animation/deblur-icu/table.qza --o-merged-table /home/yovazquezbaeza/research/ag-animation/deblur-fmt-and-icu/table.qza\n",
    "#\n",
    "# 1.3) Remove blooms\n",
    "# This script will remove the bloom sequences from the table and\n",
    "# from the representative sequences.\n",
    "./remove-blooms.py\n",
    "qiime tools import --input-path /home/yovazquezbaeza/research/ag-animation/deblur-fmt-and-icu/representative-sequences.upper.fna --output-path /home/yovazquezbaeza/research/ag-animation/deblur-fmt-and-icu/representative_sequences.qza --type 'FeatureData[Sequence]'\n",
    "#\n",
    "# 2) Grab the ITS sequences\n",
    "#\n",
    "mkdir -p sequence-files\n",
    "cp /projects/qiita_data/preprocessed_data/157_seqs.fna /home/yovazquezbaeza/research/ag-animation/sequence-files/its-seqs.fna\n",
    "#\n",
    "# 2.1) trim to 125 NTs\n",
    "trim_fasta.py -i /home/yovazquezbaeza/research/ag-animation/sequence-files/its-seqs.fna -o /home/yovazquezbaeza/research/ag-animation/sequence-files/its.seqs.125nt.fna -l 125\n",
    "#\n",
    "# 2.2) pick OTUs at 99%.\n",
    "# qsub -l mem=128gb,nodes=1:ppn=32 -l walltime=120:00:00 -e sortmerna.e -o sortmerna.o -N closed commands.sh\n",
    "pick_closed_reference_otus.py -i /home/yovazquezbaeza/research/ag-animation/sequence-files/its.seqs.125nt.fna -o /home/yovazquezbaeza/research/ag-animation/closed-ref-its  -p /home/yovazquezbaeza/research/ag-animation/sortmerna-params.txt -r /home/yovazquezbaeza/research/gg_13_8_otus/rep_set/99_otus.fasta -t /home/yovazquezbaeza/research/gg_13_8_otus/taxonomy/99_otu_taxonomy.txt\n",
    "#\n",
    "# 3) Combine all the Illumina data (FMT, ICU and AGP).\n",
    "# 3.1) Get American gut data\n",
    "#  Justine sent me an e-mail with a spreadsheet pointing to this\n",
    "#  path: /home/mcdonadt/ag-April-26-2017/otu_table_no_blooms_125nt_with_tax_min1250.biom\n",
    "#\n",
    "mkdir -p deblur-ag\n",
    "qiime tools import --input-path /home/mcdonadt/ag-April-26-2017/otu_table_no_blooms_125nt_with_tax_min1250.biom --output-path /home/yovazquezbaeza/research/ag-animation/deblur-ag/table.qza --type 'FeatureTable[Frequency]'\n",
    "# python make-representative-sequences.py\n",
    "qiime tools import --input-path /home/yovazquezbaeza/research/ag-animation/deblur-ag/representative-sequences.upper.fna --output-path /home/yovazquezbaeza/research/ag-animation/deblur-ag/representative_sequences.qza --type 'FeatureData[Sequence]'\n",
    "#\n",
    "# 3.2) Merge illumina data\n",
    "#\n",
    "qiime feature-table merge --i-table1 /home/yovazquezbaeza/research/ag-animation/deblur-fmt-and-icu/table.noblooms.qza --i-table2 /home/yovazquezbaeza/research/ag-animation/deblur-ag/table.qza --o-merged-table /home/yovazquezbaeza/research/ag-animation/deblur-ag-fmt-icu/table.qza\n",
    "#\n",
    "qiime feature-table merge-seq-data --i-data1 /home/yovazquezbaeza/research/ag-animation/deblur-ag/representative_sequences.qza --i-data2 /home/yovazquezbaeza/research/ag-animation/deblur-fmt-and-icu/representative_sequences.qza --o-merged-data /home/yovazquezbaeza/research/ag-animation/deblur-ag-fmt-icu/representative_sequences.qza\n",
    "#\n",
    "# 3.2) Re-pick OTUs from the deblurred sequences.\n",
    "#\n",
    "qiime tools export /home/yovazquezbaeza/research/ag-animation/deblur-ag-fmt-icu/representative_sequences.qza --output-dir /home/yovazquezbaeza/research/ag-animation/deblur-ag-fmt-icu/\n",
    "#\n",
    "# submit this like:\n",
    "# qsub -l mem=128gb,nodes=1:ppn=32 -l walltime=120:00:00 -e sortmerna.e -o sortmerna.o -N closed commands.2.sh\n",
    "pick_closed_reference_otus.py -i /home/yovazquezbaeza/research/ag-animation/deblur-ag-fmt-icu/dna-sequences.fasta -o /home/yovazquezbaeza/research/ag-animation/deblur-ag-fmt-icu/closed-ref/ -p /home/yovazquezbaeza/research/ag-animation/sortmerna-params.txt -r /home/yovazquezbaeza/research/gg_13_8_otus/rep_set/99_otus.fasta -t /home/yovazquezbaeza/research/gg_13_8_otus/taxonomy/99_otu_taxonomy.txt\n",
    "#\n",
    "#\n",
    "# 3.3) Re-map into an OTU table using Daniel's script\n",
    "#\n",
    "/home/yovazquezbaeza/research/ag-animation/expand.py /home/yovazquezbaeza/research/ag-animation/deblur-ag-fmt-icu/table.qza /home/yovazquezbaeza/research/ag-animation/deblur-ag-fmt-icu/closed-ref/sortmerna_picked_otus/dna-sequences_otus.txt /home/yovazquezbaeza/research/ag-animation/deblur-ag-fmt-icu/expanded-otu-table.qza\n",
    "#\n",
    "# 3.4) Import GG tree\n",
    "qiime tools import --input-path /home/yovazquezbaeza/research/gg_13_8_otus/trees/99_otus.tree --output-path /home/yovazquezbaeza/research/ag-animation/deblur-ag-fmt-icu/closed-ref/greengenes.99.qza --type 'Phylogeny[Rooted]'\n",
    "#\n",
    "# 4) Combine re-mapped OTU table and ITS OTU table.\n",
    "#\n",
    "# 4.1) Import ITS table into QIIME2\n",
    "#\n",
    "qiime tools import --input-path /home/yovazquezbaeza/research/ag-animation/closed-ref-its/otu_table.biom  --output-path /home/yovazquezbaeza/research/ag-animation/closed-ref-its/table.qza --type 'FeatureTable[Frequency]'\n",
    "#\n",
    "# 4.2) merge OTU tables\n",
    "mkdir -p remapped-ag-fmt-icu-its\n",
    "qiime feature-table merge --i-table1 /home/yovazquezbaeza/research/ag-animation/deblur-ag-fmt-icu/expanded-otu-table.qza --i-table2 /home/yovazquezbaeza/research/ag-animation/closed-ref-its/table.qza --o-merged-table /home/yovazquezbaeza/research/ag-animation/remapped-ag-fmt-icu-its/table.qza\n",
    "qiime feature-table rarefy --i-table /home/yovazquezbaeza/research/ag-animation/remapped-ag-fmt-icu-its/table.qza --p-sampling-depth 1250 --o-rarefied-table /home/yovazquezbaeza/research/ag-animation/remapped-ag-fmt-icu-its/table.even1250.qza\n",
    "#\n",
    "#\n",
    "# 5) Use Greengenes 99% to compute UniFrac\n",
    "#\n",
    "# 5.1) compute UniFrac\n",
    "# qsub -l mem=64gb,nodes=1:ppn=10 -l walltime=120:00:00 -o state-unifrac.o -e state-unifrac.e -N state -M yoshiki89@gmail.com commands.2.sh\n",
    "qiime state-unifrac unweighted --i-table /home/yovazquezbaeza/research/ag-animation/remapped-ag-fmt-icu-its/table.even1250.qza --i-phylogeny /home/yovazquezbaeza/research/ag-animation/deblur-ag-fmt-icu/closed-ref/greengenes.99.qza --p-threads 10 --o-distance-matrix /home/yovazquezbaeza/research/ag-animation/remapped-ag-fmt-icu-its/unweighted-unifrac.even1250.qza\n",
    "#\n",
    "# 5.2) Ordinate distance matrix\n",
    "qiime diversity pcoa --i-distance-matrix /home/yovazquezbaeza/research/ag-animation/remapped-ag-fmt-icu-its/unweighted-unifrac.even1250.qza  --o-pcoa /home/yovazquezbaeza/research/ag-animation/remapped-ag-fmt-icu-its/pcoa.unweighted-unifrac.even1250.qza\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  },
  "widgets": {
   "state": {},
   "version": "1.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
